src/state/experiment.js

```diff
@@ -125,6 +126,19 @@ const loadFilePairIfNeeded = id => (dispatch, getState) => {
   );
 };
 
+const pushAnswer = (experimentId, assignmentId, answer) => (dispatch, getState) => {
+  const { experiment } = getState();
+  const assigment = experiment.assignments.find(a => a.id === assignmentId);
+  const answerPayload = {
+    answer: answer,
+    duration: assigment.duration
+  }
+  return api.putAnswer(experimentId, assignmentId, answerPayload).catch(e => {
+    dispatch(addErrors(e));
+    dispatch({ type: LOAD_ERROR, error: e });
```

Comment:
it will be super weird behavior and also race condition is possible

Reply:
I copied it from other places
https://github.com/src-d/code-annotation/blob/b2e1db20a8498b372e88f9ed58b0e378eb8b405e/src/state/experiment.js#L136
https://github.com/src-d/code-annotation/blob/b2e1db20a8498b372e88f9ed58b0e378eb8b405e/src/state/experiment.js#L179
Did I miss something? Any `dispatch` maybe?

Reply:
Yes, you are missing the logic. Of messed up with how async promises work.
Let me explain the whole flow: 

A user opens page `/exp/1/1`, what happens next:
1. middleware see the change and call `selectAssigment` action
2. `selectAssigment` sets `assignmentId` and calls `loadFilePairIfNeeded` action
3. If there are no files in cache, it calls API and returns promise
4. When promise is resolved it sets file content to store or if promise is rejected it sets error
5. User choose are those files similar or different. When user clicks button it calls `mark` action.
6. `mark` action sets answer to store
> next your changes that work wrong
7. You call `pushAnswer` action, it returns promise and sets error
8. __at the same time__ (next tick) it calls `nextAssigment` which starts changes url and everything starts over.

It's very possible that next assignment will be loaded and shown to user __before__ `PUT` api call has finished. What would happen in such case?
1. User will see next assignment and start deciding answer for the next assignment
2. Then he will get error at very random time

right now we don't reset errors anywhere but if we do later it will be even worst.
Because you don't wait for the promise, another promise can remove your error.
It will look like this for user:
1. Click button
2. Next assignment loader
3. Some weird flash
4. Next assignment, but the previous answer wasn't saved on server